File: Algorithms/Rabin Karp algorithm.cpp
// Rabin-Karp algorithm in C++

#include<bits/stdc++.h>
using namespace std;
int main (){
   char txt[80], pat[80];
   int q;
   cout<<"Enter the string to be searched : ";
   cin>>txt;
   cout<<"Enter the pattern : ";
   cin>>pat;
   int d = 256;
   cout<<"Enter the prime number : ";
   cin>>q;
   int M = strlen (pat);
   int N = strlen (txt);
   int i, j;
   int p = 0;
   int t = 0;
   int h = 1;
   for (i = 0; i < M - 1; i++)
      h = (h * d) % q;
   for (i = 0; i < M; i++){
      p = (d * p + pat[i]) % q;
      t = (d * t + txt[i]) % q;
   }
   for (i = 0; i <= N - M; i++){
      if (p == t){
         for (j = 0; j < M; j++){
            if (txt[i + j] != pat[j])
            break;
         }
         if (j == M)
            cout<<"Pattern found at index : "<<i<<endl;
      }
      if (i < N - M){
         t = (d * (t - txt[i] * h) + txt[i + M]) % q;
         if (t < 0)
            t = (t + q);
      }
   }
   return 0;
}


/*
Input:
Enter the string to be searched : SUNDAY
Enter the pattern : DAY
Enter the prime number : 13

Output:
Pattern found at index : 3



The time complexity of the searching phase of the Karp-Rabin algorithm is O(mn) (when searching for am in an for instance).
Its expected number of text character comparisons is O(n+m)

*/
  
 

File: Algorithms/rabin_karp_algorithm.py

'''

RABIN KARP ALGORITHM :

Rabin Karp algorithm matches the hash value of the pattern with the hash value of current substring of text, and if the hash values match then only it starts matching individual characters

# first test case :
Text = "Hello LGM SOC participant"
Pattern = "LGM"
Output -> pattern found at index 6 in the text

# second test case :
Text = "123 3422 1 3 2 2 12333"
Pattern = "23"
Output -> Pattern found at index 1, and 18 in the text

'''


# Any prime number taken
prime = 1000000007

# number of possible characters in the input alphabet
d = 256

# variable defined to check if pattern is found in text or not
found = 0

# function defined for Rabin Karp Algorithm
def rabin_karp_algorithm(pattern, text):
	global found

	# arbitrary variable j and h taken
	j = 0
	h = 1
	pattern_hash_value = 0 # variable taken to store pattern hash value
	text_hash_value = 0 # variable taken to store text hash value

	# variable h is defined as (d^(M-1))%prime
	for i in range(len(pattern)-1):
		h = (h*d)%prime

	# Calculating the hash value of pattern and text and storing in pattern_hash_value and text_hash_value respectively
	for i in range(len(pattern)):
		pattern_hash_value = (d*pattern_hash_value + ord(pattern[i]))%prime
		text_hash_value = (d*text_hash_value + ord(text[i]))%prime

	# Searching the pattern over text one by one, and if found, then we check pattern character one by one
	for i in range(len(text)-len(pattern)+1):
		if pattern_hash_value==text_hash_value:
			# Check for characters one by one as here hash_value matched
			for j in range(len(pattern)):
				if text[i+j] != pattern[j]:
					break
				else:
					j+=1 # updating the j accordingly

			# j becomes len(pattern), then we print that pattern found at i index
			if j==len(pattern):
				found = 1
				print("Pattern found at index " + str(i))

		# Calculating the  hash value for next window of text bu removing leading digit, add trailing digit
		if i < len(text)-len(pattern):
			text_hash_value = (d*(text_hash_value-ord(text[i])*h) + ord(text[i+len(pattern)]))%prime

			# if we get negative hashvalue, then we convert it to positive by adding prime to it
			if text_hash_value < 0:
				text_hash_value = text_hash_value+prime

	# if pattern is not found then we print this message
	if(found == 0):
		print("Pattern is not present in Text.")



# input
text = "Hello LGM SOC participant"
pattern = "Z"

print("\n# INPUT : ")
print("Text : ", text)
print("Pattern : ", pattern)
print("\n# OUTPUT : ")
# main function called
rabin_karp_algorithm(pattern,text)
